OpenSearch Distance Calculation Feature

mmiguerodriguez opened the pull request:
This branch extends EvoMaster with distance calculation support for OpenSearch queries.

### Features
- Added OPENSEARCH to Heuristic DTO Entry type.

**Distance Data Structures**
- OpenSearchDistanceWithMetrics: stores distance and number of evaluated documents (non-negative enforced).
- OpenSearchCommandWithDistance: pairs commands with their computed distance.

**OpenSearchHandler**
- Captures OpenSearch queries during execution.
- Retrieves documents from indices.
- Computes query–document distances and records minimum distance + evaluated count.
- Term Query distance calculation (integers).

### Testing
- Example Age app demonstrating equals and range logic.

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/internal/SutController.java:
```diff
@@ -323,6 +327,17 @@ public final void initMongoHandler() {
         }
     }
 
+    public final void initOpenSearchHandler() {
```
it is fine in this PR. but, as this code seems very similar to what done for Mongo, maybe @jgaleotti  you can refactor it to avoid duplication? 
@mmiguerodriguez   maybe leave a `// TODO` comment stating this code likely needs refactoring with the Mongo one

jgaleotti replied:
please leve a TODO and when the Redis and the OpenSearch implementations are done I will refactor them out.

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/internal/SutController.java:
```diff
@@ -467,6 +490,34 @@ public final void computeMongoHeuristics(ExtraHeuristicsDto dto, List<Additional
+                    try {
+                        openSearchHandler.handle(it);
+                    } catch (Exception e){
+                        SimpleLogger.error("FAILED TO HANDLE OPENSEARCH COMMAND");
```
add info on the error in this log, eg, `+ e.getMssage()`

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/internal/db/OpenSearchDistanceWithMetrics.java:
```diff
@@ -0,0 +1,27 @@
+
+    public final double distance;
+
+    public final int numberOfEvaluatedDocuments;
```
those fields should be private, as anyway you are providing public getters

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/opensearch/OpenSearchHeuristicsCalculator.java:
```diff
@@ -0,0 +1,111 @@
+            return calculateDistanceForEquals((TermOperation<?>) operation, doc);
+        }
+
+        return 0;
```
is it correct to return 0 here? if so, it is not so obvious. add some /* */ comments to clarify it

jgaleotti replied:
I guess this should be Double.MAX_VALUE as this is the case where it is not supported, right?

mmiguerodriguez replied:
Yes, will modify this.

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/opensearch/operations/ComparisonOperation.java:
```diff
@@ -0,0 +1,19 @@
+    private final String fieldName;
+    private final V value;
+
+    ComparisonOperation(String fieldName, V value) {
```
default-scope in Java is very tricky, and possibly counter-intuitive. as such, I usually avoid it. consider if a `protected` makes sense here

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/opensearch/selectors/QuerySelector.java:
```diff
@@ -0,0 +1,37 @@
+    public abstract QueryOperation getOperation(Object query);
+
+    protected Boolean isUniqueEntry(Map<?, ?> map) {
+        return map.size() == 1;
```
what is input map is null?

mmiguerodriguez replied:
Removed this method since unused

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/opensearch/utils/OpenSearchQueryHelper.java:
```diff
@@ -0,0 +1,88 @@
+            Object kind = value.getClass().getMethod("_kind").invoke(value);
+            String kindName = (String) kind.getClass().getMethod("name").invoke(kind);
+
+            switch (kindName) {
```
what about `Integer` and `Short`?

mmiguerodriguez replied:
I used OpenSearch [FieldValue](https://github.com/opensearch-project/opensearch-java/blob/main/java-client/src/main/java/org/opensearch/client/opensearch/_types/FieldValue.java#L74) as reference. Will review if more are required.

arcuri82 commented on client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/OpenSearchCommand.java:
```diff
@@ -1,12 +1,13 @@
      * Name of the index that the operation was applied to
      */
-    private final Object index;
+    private final List<String> index;
```
update docs based on new type. ie, how can a list of strings be a name?

mmiguerodriguez replied:
Updated docs following OpenSearch SearchRequest docs.
```
    /**
     * A comma-separated list of index names to search; use <code>_all</code> or
     * empty string to perform the operation on all indices
     * <p>
     * API name: {@code index}
     */
```

arcuri82 commented on client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/OpenSearchClientClassReplacement.java:
```diff
@@ -71,13 +73,32 @@ private static Method retrieveMethod(Object openSearchClient, String id) {
+            } else if (result instanceof String) {
+                return Collections.singletonList((String) result);
+            } else {
+                return Collections.singletonList(result.toString());
```
is this last case expected in normal behaviour? or is it just a safety check to not crash EM? in this latter case, might want to log such event (so we can see when/if it happens, and fix it if needed)

mmiguerodriguez replied:
I made it this way to consider all cases, but the normal behavior would be for the index to be a single String.

mmiguerodriguez replied:
Added logs

arcuri82 commented on e2e-tests/spring-rest-opensearch/src/main/java/com/opensearch/age/OpenSearchAgeRest.java:
```diff
@@ -0,0 +1,69 @@
+        return ageRepo.findByAge(age);
+    }
+
+//    @GetMapping("gte/{gte}")
```
is this code no longer needed? or to be put back in future? either remove it, or add a TODO comment if you ll need it in future

mmiguerodriguez replied:
Will be put back when testing other types of queries (eg: Range). Will add TODO comment.

arcuri82 commented on e2e-tests/spring-rest-opensearch/src/test/java/org/evomaster/e2etests/spring/rest/opensearch/age/OpenSearchAgeEMTest.java:
```diff
@@ -0,0 +1,35 @@
+
+                Solution<RestIndividual> solution = initAndRun(args);
+
+//                assertHasAtLeastOne(solution, HttpVerb.GET, 404, "/age/{q}", null);
```
why no assertions? if the techniques work, when you should be able to cover the different endpoints you have designed. if not yet, and tehcniques needs more fixing, it is ok to have a PR now, but: (1) add the needed assertions on the different endpoints; (2) mark the test with `@Disabled`; (3) leave a TODO comment stating this is failing and need more work

arcuri82 commented on client-java/controller/src/main/java/org/evomaster/client/java/controller/opensearch/utils/OpenSearchQueryHelper.java:
```diff
@@ -0,0 +1,90 @@
+                    return null;
+            }
+
+            SimpleLogger.debug("[OpenSearch] [extractTypedFieldValue] found unknown kind: " + kindName);
```
should be a `warn` not `debug`

arcuri82 commented on client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/OpenSearchClientClassReplacement.java:
```diff
@@ -71,13 +74,34 @@ private static Method retrieveMethod(Object openSearchClient, String id) {
+                SimpleLogger.debug("[OpenSearch] [getIndex] returned list of length: " + ((List<?>) result).size());
+                return (List<String>) result;
+            } else {
+                SimpleLogger.debug("[OpenSearch] [getIndex] returned unknown type: " + result.getClass().getName());
```
should be a `warn` not `debug`

mmiguerodriguez commented:
I've addressed all of the comments. Regarding the question from https://github.com/WebFuzzing/EvoMaster/pull/1301#discussion_r2310769393, at least for filtering those are the minimal types that should be mapped.
